### PR TITLE
Blood Type Refactor

### PR DESCRIPTION
**Pull request by nullbear**
not really any reason to rush this at the moment. But i'm touching blood
stuff anyways.

This proc is more reuseable, as it returns a list of blood types
potentially acceptable for transfusion instead of true/false, and only
takes a single argument (which means if something is checking blood type
compatibility, it only needs the type, rather than the species and blood
type of the donator etcetera, and could be used for stuff like scanners
and junk.)

It's also a bit more readable, making it easier to add new blood types
for xeno races, if ever needed.

Should also fix inverted blood transfusion math that people were experiencing.

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Neat

**Comment by nullbear** [comment]
wtf some weird tab issue with gh

File: code/modules/mob/living/carbon/human/blood.dm [hunk 1]
```diff
@@ -245,28 +245,28 @@ var/const/BLOOD_VOLUME_SURVIVE = 122
...
+		if("B+")
+			return list("B-", "B+", "O-", "O+")
+		if("AB-")
+			return list("A-", "B-", "O-", "AB-")
+		if("AB+")
+			return list("A-", "A+", "B-", "B+", "O-", "O+", "AB-", "AB+")
+		if("O-")
+			return list("O-")
+		if("O+")
+			return list("O+")
```
Comment (KazeEspada) [inline code/modules/mob/living/carbon/human/blood.dm]: Shouldnt O- be on this one too?

**Comment by Iamgoofball** [comment]
datum blood types when

File: code/modules/mob/living/carbon/human/blood.dm (hunk 1 above)
Comment (nullbear) [inline code/modules/mob/living/carbon/human/blood.dm]: no

**Comment by optimumtact** [comment]
Not much point I would say goof, unless you want to invent some new blood types for some reason

**Comment by Shapsy** [comment]
C type blood when

**Comment by Zilenan91** [comment]
There's still not enough blood packs in Box medbay

**Comment by Iamgoofball** [comment]
@optimumtact itd be a little cleaner

**Comment by nullbear** [comment]
i don't see the point to datum 'blood types' at the moment, since it's essentially just a 2 character string that is only used like twice ever. If you mean datum 'blood' that would be nice.

**Comment by nullbear** [comment]
And i was gunna add an 'X' bloodtype, for xenos, which happens to be acidic.

**Comment by coiax** [comment]
What about ponies? Is there more than one kind of liquid friendship?

**Comment by octareenroon91** [comment]
Do xenos not have X* blood type anymore?

**Comment by nullbear** [comment]
nah that's bay.

File: code/modules/mob/living/carbon/human/blood.dm [hunk 2]
```diff
@@ -226,7 +226,7 @@ var/const/BLOOD_VOLUME_SURVIVE = 122
 	if (!injected || !our || !dna)
 		return
 
-	if(blood_incompatible(injected.data["blood_type"], dna.blood_type,injected.data["species"], dna.species.id) )
+	if (!(injected.data["blood_type"] in get_safe_blood(dna.blood_type)) && injected.data["species"] != dna.species.id)
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/blood.dm]: "||" not "&&"

**Comment by octareenroon91** [comment]
Assuredly, there are mentions of X* in our codebase.

File: code/modules/mob/living/carbon/human/blood.dm (hunk 2 above)
Comment (nullbear) [inline code/modules/mob/living/carbon/human/blood.dm]: oop, i thought i changed that.

**Pull request merged**